File: src/audit.ts
import { access, readFile, readdir } from "node:fs/promises";
import { dirname, join } from "node:path";
import type { DiagnosticBag } from "./diagnostics.js";
import { assetFile, normalizeModelLocation, normalizeTextureLocation, splitLocation } from "./resource-location.js";
import { isObject, type JsonObject, type JsonValue } from "./types.js";

export interface AuditInput {
  resourceRoot: string;
  items: JsonObject;
  blocks: JsonObject;
  images?: JsonObject;
  blueprintRoot?: string;
}

export interface AuditSummary {
  referencedModels: number;
  resolvedModels: number;
  generatedModels: number;
  referencedBlueprints: number;
  missingBlueprints: number;
  copiedItemDefinitions: number;
  referencedTextures: number;
  resolvedTextures: number;
  missingModels: number;
  missingTextures: number;
}

async function exists(file: string): Promise<boolean> {
  try { await access(file); return true; } catch { return false; }
}

async function listJsonFiles(directory: string): Promise<string[]> {
  if (!(await exists(directory))) return [];
  const files: string[] = [];
  const visit = async (path: string): Promise<void> => {
    for (const entry of await readdir(path, { withFileTypes: true })) {
      const child = join(path, entry.name);
      if (entry.isDirectory()) await visit(child);
      else if (entry.isFile() && entry.name.toLowerCase().endsWith(".json")) files.push(child);
    }
  };
  await visit(directory);
  return files;
}

function knownVanillaParent(location: string): boolean {
  if (location === "builtin:entity" || location.startsWith("builtin:")) return true;
  if (!location.startsWith("minecraft:")) return false;
  const path = location.slice("minecraft:".length);
  return path === "item/generated" || path === "item/handheld" || path === "item/handheld_rod" || path.startsWith("item/template_") ||
    path === "block/block" || path === "block/cube" || path.startsWith("block/cube_") || path.startsWith("block/orientable") ||
    path === "block/cross" || path.startsWith("block/template_") || path.includes("stairs") || path.includes("slab");
}

function addModel(refs: Map<string, boolean>, raw: string, generated: boolean): void {
  refs.set(raw, (refs.get(raw) ?? false) || generated);
}

function collectGeneration(generation: JsonObject, modelRefs: Map<string, boolean>, textureRefs: Set<string>): void {
  if (typeof generation.parent === "string") addModel(modelRefs, generation.parent, false);
  if (isObject(generation.textures)) for (const texture of Object.values(generation.textures)) {
    if (typeof texture === "string" && !texture.startsWith("#")) textureRefs.add(texture);
  }
}

function collectModelNodes(value: JsonValue, modelRefs: Map<string, boolean>, textureRefs: Set<string>, blueprintRefs: Set<string>, parentKey = ""): void {
  if (Array.isArray(value)) {
    for (const entry of value) collectModelNodes(entry, modelRefs, textureRefs, blueprintRefs, parentKey);
    return;
  }
  if (!isObject(value)) {
    if (typeof value === "string" && parentKey === "model") addModel(modelRefs, value, false);
    return;
  }
  const type = typeof value.type === "string" ? value.type.replace(/^minecraft:/, "") : "";
  const generation = isObject(value.generation) ? value.generation : undefined;
  if (type === "model") {
    const path = typeof value.path === "string" ? value.path : typeof value.model === "string" ? value.model : undefined;
    if (path) addModel(modelRefs, path, Boolean(generation) || typeof value.blueprint === "string");
  } else if (type === "special" && typeof value.base === "string") {
    addModel(modelRefs, value.base, Boolean(generation) || typeof value.blueprint === "string");
  } else if (typeof value.path === "string" && ("predicate" in value || generation)) {
    addModel(modelRefs, value.path, Boolean(generation) || typeof value.blueprint === "string");
  }
  if (generation) collectGeneration(generation, modelRefs, textureRefs);
  if (typeof value.blueprint === "string") blueprintRefs.add(value.blueprint);
  for (const [key, entry] of Object.entries(value)) collectModelNodes(entry, modelRefs, textureRefs, blueprintRefs, key);
}

async function readObject(file: string, diagnostics: DiagnosticBag, code: string): Promise<JsonObject | undefined> {
  try {
    const raw = JSON.parse((await readFile(file, "utf8")).replace(/^\uFEFF/, "")) as JsonValue;
    if (!isObject(raw)) throw new Error("JSON root is not an object");
    return raw;
  } catch (error) {
    diagnostics.error(code, String(error), { source: file });
    return undefined;
  }
}

export async function auditResourceGraph(input: AuditInput, diagnostics: DiagnosticBag): Promise<AuditSummary> {
  const modelRefs = new Map<string, boolean>();
  const textureRefs = new Set<string>();
  const blueprintRefs = new Set<string>();
  const generatedPointers = new Set<string>();
  for (const item of Object.values(input.items)) if (isObject(item)) {
    if (item.model !== undefined) collectModelNodes(item.model, modelRefs, textureRefs, blueprintRefs);
    if (item.legacy_model !== undefined) collectModelNodes(item.legacy_model, modelRefs, textureRefs, blueprintRefs);
    if (typeof item.item_model === "string" && item.model !== undefined) generatedPointers.add(item.item_model);
  }
  for (const block of Object.values(input.blocks)) if (isObject(block)) collectModelNodes(block, modelRefs, textureRefs, blueprintRefs);
  for (const image of Object.values(input.images ?? {})) if (isObject(image) && typeof image.file === "string") textureRefs.add(image.file);

  let missingBlueprints = 0;
  const blueprintRoot = input.blueprintRoot ?? join(dirname(input.resourceRoot), "blueprint");
  for (const blueprint of blueprintRefs) {
    const file = join(blueprintRoot, (blueprint.endsWith(".bbmodel") ? blueprint : blueprint + ".bbmodel").replaceAll("/", "\\"));
    if (!(await exists(file))) {
      diagnostics.error("BLUEPRINT_FILE_MISSING", "Referenced Blockbench blueprint does not exist: " + blueprint, { source: file, lossy: true });
      missingBlueprints++;
    }
  }
  let copiedItemDefinitions = 0;
  const assetsRoot = join(input.resourceRoot, "assets");
  for (const file of await listJsonFiles(assetsRoot)) {
    const normalized = file.replaceAll("\\", "/");
    if (!normalized.includes("/items/")) continue;
    copiedItemDefinitions++;
    const definition = await readObject(file, diagnostics, "ITEM_DEFINITION_JSON_INVALID");
    if (definition?.model !== undefined) collectModelNodes(definition.model, modelRefs, textureRefs, blueprintRefs);
  }
  for (const pointer of generatedPointers) {
    const [namespace, path] = splitLocation(pointer.includes(":") ? pointer : "minecraft:" + pointer);
    const file = join(input.resourceRoot, "assets", namespace, "items", path + ".json");
    if (await exists(file)) diagnostics.warning("ITEM_DEFINITION_CONFLICT", "A copied item definition occupies a path CraftEngine will generate: " + pointer, { source: file, lossy: true });
  }

  const visitedModels = new Set<string>();
  let generatedModels = 0;
  let missingModels = 0;
  let missingTextures = 0;
  const visitModel = async (rawLocation: string, generated: boolean, source: string): Promise<void> => {
    if (rawLocation.startsWith("builtin/")) return;
    const location = rawLocation.startsWith("builtin:") ? rawLocation : normalizeModelLocation(rawLocation, diagnostics, { source, field: "resource graph" });
    if (!location || visitedModels.has(location)) return;
    visitedModels.add(location);
    if (generated) {
      generatedModels++;
      return;
    }
    if (knownVanillaParent(location)) return;
    const file = assetFile(input.resourceRoot, "models", location, ".json");
    if (!(await exists(file))) {
      diagnostics.error("MODEL_FILE_MISSING", "Referenced static model does not exist: " + location, { source, field: file, lossy: true });
      missingModels++;
      return;
    }
    const model = await readObject(file, diagnostics, "MODEL_JSON_INVALID");
    if (!model) return;
    if (typeof model.parent === "string") await visitModel(model.parent, false, file);
    if (isObject(model.textures)) for (const texture of Object.values(model.textures)) {
      if (typeof texture !== "string" || texture.startsWith("#")) continue;
      textureRefs.add(texture);
    }
    if (Array.isArray(model.overrides)) for (const override of model.overrides) {
      if (isObject(override) && typeof override.model === "string") await visitModel(override.model, false, file);
    }
  };

  for (const [location, generated] of modelRefs) await visitModel(location, generated, "configuration");
  const normalizedTextures = new Set<string>();
  for (const texture of textureRefs) {
    const location = normalizeTextureLocation(texture, diagnostics, { source: "resource graph", field: "texture" });
    if (!location || location === "minecraft:missingno" || normalizedTextures.has(location)) continue;
    normalizedTextures.add(location);
    const file = assetFile(input.resourceRoot, "textures", location, ".png");
    if (!(await exists(file))) {
      diagnostics.error("TEXTURE_FILE_MISSING", "Referenced texture does not exist: " + location, { source: file, lossy: true });
      missingTextures++;
    }
  }
  return {
    referencedModels: modelRefs.size,
    resolvedModels: visitedModels.size - missingModels,
    generatedModels,
    referencedBlueprints: blueprintRefs.size,
    missingBlueprints,
    copiedItemDefinitions,
    referencedTextures: normalizedTextures.size,
    resolvedTextures: normalizedTextures.size - missingTextures,
    missingModels,
    missingTextures,
  };
}

File: src/audit.rs
//! Resource-graph audit: resolves model/texture/blueprint references from the
//! item/block configuration against the files of the generated resource pack.
//!
//! Port of `legacy/src/audit.ts`. Diagnostic codes, messages, counting
//! semantics, and reference ordering (JS Map/Set insertion order) are kept
//! identical to the TypeScript original.

use std::collections::{HashMap, HashSet};
use std::path::Path;

use serde_json::Value;

use crate::diagnostics::{Details, DiagnosticBag};
use crate::json::JsonObject;
use crate::resource_location::{
    asset_file, normalize_model_location, normalize_texture_location, split_location,
    ASSET_MODELS, ASSET_TEXTURES,
};

pub struct AuditInput<'a> {
    pub resource_root: String,
    pub items: &'a JsonObject,
    pub blocks: &'a JsonObject,
    pub images: Option<&'a JsonObject>,
    pub blueprint_root: Option<String>,
}

#[derive(Debug, Clone)]
pub struct AuditSummary {
    pub referenced_models: usize,
    pub resolved_models: usize,
    pub generated_models: usize,
    pub referenced_blueprints: usize,
    pub missing_blueprints: usize,
    pub copied_item_definitions: usize,
    pub referenced_textures: usize,
    pub resolved_textures: usize,
    pub missing_models: usize,
    pub missing_textures: usize,
}

/// Insertion-ordered string set (JS Set semantics: first-insert position, dedup).
#[derive(Default)]
struct OrderedSet {
    order: Vec<String>,
    seen: HashSet<String>,
}

impl OrderedSet {
    /// Returns true when the value was newly inserted.
    fn insert(&mut self, value: &str) -> bool {
        if self.seen.insert(value.to_string()) {
            self.order.push(value.to_string());
            true
        } else {
            false
        }
    }

    fn len(&self) -> usize {
        self.order.len()
    }

    fn iter(&self) -> std::slice::Iter<'_, String> {
        self.order.iter()
    }
}

/// Insertion-ordered model reference map (JS Map semantics); the generated
/// flag is OR-merged on duplicate keys.
#[derive(Default)]
struct ModelRefs {
    entries: Vec<(String, bool)>,
    index: HashMap<String, usize>,
}

impl ModelRefs {
    fn add(&mut self, raw: &str, generated: bool) {
        match self.index.get(raw) {
            Some(&slot) => self.entries[slot].1 |= generated,
            None => {
                self.index.insert(raw.to_string(), self.entries.len());
                self.entries.push((raw.to_string(), generated));
            }
        }
    }

    fn len(&self) -> usize {
        self.entries.len()
    }

    fn iter(&self) -> std::slice::Iter<'_, (String, bool)> {
        self.entries.iter()
    }
}

fn exists(path: &str) -> bool {
    Path::new(path).exists()
}

/// Node `dirname` equivalent.
fn dirname(path: &str) -> String {
    match Path::new(path).parent() {
        Some(parent) if !parent.as_os_str().is_empty() => parent.to_string_lossy().into_owned(),
        _ => ".".to_string(),
    }
}

/// Recursively list `.json` files (case-insensitive extension), like the TS
/// `listJsonFiles` walker. Missing directory yields an empty list.
fn list_json_files(directory: &Path) -> Vec<std::path::PathBuf> {
    if !directory.exists() {
        return Vec::new();
    }
    let mut files = Vec::new();
    for entry in walkdir::WalkDir::new(directory)
        .into_iter()
        .filter_map(|entry| entry.ok())
    {
        let name = entry.file_name().to_string_lossy();
        if entry.file_type().is_file() && name.to_lowercase().ends_with(".json") {
            files.push(entry.into_path());
        }
    }
    files
}

fn known_vanilla_parent(location: &str) -> bool {
    if location.starts_with("builtin:") {
        return true;
    }
    let Some(path) = location.strip_prefix("minecraft:") else {
        return false;
    };
    path == "item/generated"
        || path == "item/handheld"
        || path == "item/handheld_rod"
        || path.starts_with("item/template_")
        || path == "block/block"
        || path == "block/cube"
        || path.starts_with("block/cube_")
        || path.starts_with("block/orientable")
        || path == "block/cross"
        || path.starts_with("block/template_")
        || path.contains("stairs")
        || path.contains("slab")
}

fn collect_generation(
    generation: &JsonObject,
    model_refs: &mut ModelRefs,
    texture_refs: &mut OrderedSet,
) {
    if let Some(parent) = generation.get("parent").and_then(Value::as_str) {
        model_refs.add(parent, false);
    }
    if let Some(textures) = generation.get("textures").and_then(Value::as_object) {
        for texture in textures.values() {
            if let Some(texture) = texture.as_str() {
                if !texture.starts_with('#') {
                    texture_refs.insert(texture);
                }
            }
        }
    }
}

fn collect_model_nodes(
    value: &Value,
    model_refs: &mut ModelRefs,
    texture_refs: &mut OrderedSet,
    blueprint_refs: &mut OrderedSet,
    parent_key: &str,
) {
    if let Some(entries) = value.as_array() {
        for entry in entries {
            collect_model_nodes(entry, model_refs, texture_refs, blueprint_refs, parent_key);
        }
        return;
    }
    let object = match value.as_object() {
        Some(object) => object,
        None => {
            // Bare strings under a "model" key are plain model references.
            if let Some(raw) = value.as_str() {
                if parent_key == "model" {
                    model_refs.add(raw, false);
                }
            }
            return;
        }
    };
    let node_type = match object.get("type").and_then(Value::as_str) {
        Some(value) => value.strip_prefix("minecraft:").unwrap_or(value),
        None => "",
    };
    let generation = object.get("generation").and_then(Value::as_object);
    let blueprint = object.get("blueprint").and_then(Value::as_str);
    let generated = generation.is_some() || blueprint.is_some();
    let base = object.get("base").and_then(Value::as_str);
    if node_type == "model" {
        let path = object
            .get("path")
            .and_then(Value::as_str)
            .or_else(|| object.get("model").and_then(Value::as_str));
        if let Some(path) = path {
            model_refs.add(path, generated);
        }
    } else if node_type == "special" && base.is_some() {
        model_refs.add(base.unwrap(), generated);
    } else if object.contains_key("predicate") || generation.is_some() {
        if let Some(path) = object.get("path").and_then(Value::as_str) {
            model_refs.add(path, generated);
        }
    }
    if let Some(generation) = generation {
        collect_generation(generation, model_refs, texture_refs);
    }
    if let Some(blueprint) = blueprint {
        blueprint_refs.insert(blueprint);
    }
    for (key, entry) in object {
        collect_model_nodes(entry, model_refs, texture_refs, blueprint_refs, key);
    }
}

/// Read a JSON object file, reporting `code` diagnostics on failure.
/// Strips a leading UTF-8 BOM like the TS reader.
fn read_object(file: &str, diagnostics: &mut DiagnosticBag, code: &str) -> Option<JsonObject> {
    let content = match std::fs::read_to_string(file) {
        Ok(content) => content,
        Err(error) => {
            diagnostics.error(code, &error.to_string(), Details::new().source(file));
            return None;
        }
    };
    let content = content.strip_prefix('\u{FEFF}').unwrap_or(content.as_str());
    match serde_json::from_str::<Value>(content) {
        Ok(Value::Object(object)) => Some(object),
        Ok(_) => {
            diagnostics.error(code, "JSON root is not an object", Details::new().source(file));
            None
        }
        Err(error) => {
            diagnostics.error(code, &error.to_string(), Details::new().source(file));
            None
        }
    }
}

#[allow(clippy::too_many_arguments)]
fn visit_model(
    raw_location: &str,
    generated: bool,
    source: &str,
    input: &AuditInput<'_>,
    diagnostics: &mut DiagnosticBag,
    visited_models: &mut OrderedSet,
    texture_refs: &mut OrderedSet,
    generated_models: &mut usize,
    missing_models: &mut usize,
) {
    if raw_location.starts_with("builtin/") {
        return;
    }
    let location = if raw_location.starts_with("builtin:") {
        raw_location.to_string()
    } else {
        match normalize_model_location(
            raw_location,
            diagnostics,
            &Details::new().source(source).field("resource graph"),
        ) {
            Some(location) => location,
            None => return,
        }
    };
    if !visited_models.insert(&location) {
        return;
    }
    if generated {
        *generated_models += 1;
        return;
    }
    if known_vanilla_parent(&location) {
        return;
    }
    let file = asset_file(&input.resource_root, ASSET_MODELS, &location, ".json");
    if !exists(&file) {
        diagnostics.error(
            "MODEL_FILE_MISSING",
            &format!("Referenced static model does not exist: {}", location),
            Details::new().source(source).field(file).lossy(),
        );
        *missing_models += 1;
        return;
    }
    let model = match read_object(&file, diagnostics, "MODEL_JSON_INVALID") {
        Some(model) => model,
        None => return,
    };
    if let Some(parent) = model.get("parent").and_then(Value::as_str) {
        let parent = parent.to_string();
        visit_model(
            &parent,
            false,
            &file,
            input,
            diagnostics,
            visited_models,
            texture_refs,
            generated_models,
            missing_models,
        );
    }
    if let Some(textures) = model.get("textures").and_then(Value::as_object) {
        for texture in textures.values() {
            if let Some(texture) = texture.as_str() {
                if !texture.starts_with('#') {
                    texture_refs.insert(texture);
                }
            }
        }
    }
    if let Some(overrides) = model.get("overrides").and_then(Value::as_array) {
        for override_entry in overrides {
            if let Some(target) = override_entry.get("model").and_then(Value::as_str) {
                let target = target.to_string();
                visit_model(
                    &target,
                    false,
                    &file,
                    input,
                    diagnostics,
                    visited_models,
                    texture_refs,
                    generated_models,
                    missing_models,
                );
            }
        }
    }
}

pub fn audit_resource_graph(input: &AuditInput<'_>, diagnostics: &mut DiagnosticBag) -> AuditSummary {
    let mut model_refs = ModelRefs::default();
    let mut texture_refs = OrderedSet::default();
    let mut blueprint_refs = OrderedSet::default();
    let mut generated_pointers = OrderedSet::default();

    for item in input.items.values() {
        let Some(item) = item.as_object() else { continue };
        if let Some(model) = item.get("model") {
            collect_model_nodes(model, &mut model_refs, &mut texture_refs, &mut blueprint_refs, "");
        }
        if let Some(legacy_model) = item.get("legacy_model") {
            collect_model_nodes(legacy_model, &mut model_refs, &mut texture_refs, &mut blueprint_refs, "");
        }
        if item.contains_key("model") {
            if let Some(item_model) = item.get("item_model").and_then(Value::as_str) {
                generated_pointers.insert(item_model);
            }
        }
    }
    for block in input.blocks.values() {
        if block.is_object() {
            collect_model_nodes(block, &mut model_refs, &mut texture_refs, &mut blueprint_refs, "");
        }
    }
    if let Some(images) = input.images {
        for image in images.values() {
            if let Some(file) = image.get("file").and_then(Value::as_str) {
                texture_refs.insert(file);
            }
        }
    }

    // Blueprint existence runs before the item-definition scan, exactly like
    // the TS order: blueprints first seen in copied definitions are counted
    // but never checked.
    let blueprint_root = match &input.blueprint_root {
        Some(root) => root.clone(),
        None => Path::new(&dirname(&input.resource_root))
            .join("blueprint")
            .to_string_lossy()
            .into_owned(),
    };
    let mut missing_blueprints = 0usize;
    for blueprint in blueprint_refs.iter() {
        let name = if blueprint.ends_with(".bbmodel") {
            blueprint.clone()
        } else {
            format!("{}.bbmodel", blueprint)
        };
        // TS rewrites "/" to "\\" before join (Windows blueprint layout).
        let file = Path::new(&blueprint_root)
            .join(name.replace('/', "\\"))
            .to_string_lossy()
            .into_owned();
        if !exists(&file) {
            diagnostics.error(
                "BLUEPRINT_FILE_MISSING",
                &format!("Referenced Blockbench blueprint does not exist: {}", blueprint),
                Details::new().source(file).lossy(),
            );
            missing_blueprints += 1;
        }
    }

    let mut copied_item_definitions = 0usize;
    let assets_root = Path::new(&input.resource_root).join("assets");
    for file in list_json_files(&assets_root) {
        let file = file.to_string_lossy().into_owned();
        // TS slash-normalizes the walked path before the "/items/" check.
        if !file.replace('\\', "/").contains("/items/") {
            continue;
        }
        copied_item_definitions += 1;
        if let Some(definition) = read_object(&file, diagnostics, "ITEM_DEFINITION_JSON_INVALID") {
            if let Some(model) = definition.get("model") {
                collect_model_nodes(model, &mut model_refs, &mut texture_refs, &mut blueprint_refs, "");
            }
        }
    }
    for pointer in generated_pointers.iter() {
        let qualified = if pointer.contains(':') {
            pointer.clone()
        } else {
            format!("minecraft:{}", pointer)
        };
        let (namespace, path) = split_location(&qualified);
        let file = Path::new(&input.resource_root)
            .join("assets")
            .join(namespace)
            .join("items")
            .join(format!("{}.json", path))
            .to_string_lossy()
            .into_owned();
        if exists(&file) {
            diagnostics.warning(
                "ITEM_DEFINITION_CONFLICT",
                &format!(
                    "A copied item definition occupies a path CraftEngine will generate: {}",
                    pointer
                ),
                Details::new().source(file).lossy(),
            );
        }
    }

    let mut visited_models = OrderedSet::default();
    let mut generated_models = 0usize;
    let mut missing_models = 0usize;
    for (location, generated) in model_refs.iter() {
        visit_model(
            location,
            *generated,
            "configuration",
            input,
            diagnostics,
            &mut visited_models,
            &mut texture_refs,
            &mut generated_models,
            &mut missing_models,
        );
    }

    let mut normalized_textures = OrderedSet::default();
    let mut missing_textures = 0usize;
    for texture in texture_refs.iter() {
        let location = match normalize_texture_location(
            texture,
            diagnostics,
            &Details::new().source("resource graph").field("texture"),
        ) {
            Some(location) => location,
            None => continue,
        };
        if location == "minecraft:missingno" || !normalized_textures.insert(&location) {
            continue;
        }
        let file = asset_file(&input.resource_root, ASSET_TEXTURES, &location, ".png");
        if !exists(&file) {
            diagnostics.error(
                "TEXTURE_FILE_MISSING",
                &format!("Referenced texture does not exist: {}", location),
                Details::new().source(file).lossy(),
            );
            missing_textures += 1;
        }
    }

    AuditSummary {
        referenced_models: model_refs.len(),
        resolved_models: visited_models.len() - missing_models,
        generated_models,
        referenced_blueprints: blueprint_refs.len(),
        missing_blueprints,
        copied_item_definitions,
        referenced_textures: normalized_textures.len(),
        resolved_textures: normalized_textures.len() - missing_textures,
        missing_models,
        missing_textures,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn known_vanilla_parents_match_ts() {
        assert!(known_vanilla_parent("builtin:entity"));
        assert!(known_vanilla_parent("minecraft:item/generated"));
        assert!(known_vanilla_parent("minecraft:block/template_torch"));
        assert!(known_vanilla_parent("minecraft:block/oak_stairs"));
        assert!(!known_vanilla_parent("minecraft:item/diamond"));
        assert!(!known_vanilla_parent("custom:block/cube"));
    }

    #[test]
    fn collects_model_texture_and_blueprint_refs() {
        let value: Value = json!({
            "type": "minecraft:model",
            "path": "custom:item/sword",
            "blueprint": "models/sword",
            "generation": {
                "parent": "minecraft:item/generated",
                "textures": { "layer0": "#0", "side": "custom:block/steel" }
            },
            "variants": [{ "model": "custom:item/sword_guard" }]
        });
        let mut model_refs = ModelRefs::default();
        let mut texture_refs = OrderedSet::default();
        let mut blueprint_refs = OrderedSet::default();
        collect_model_nodes(&value, &mut model_refs, &mut texture_refs, &mut blueprint_refs, "");
        let models: Vec<(&str, bool)> = model_refs
            .iter()
            .map(|(raw, generated)| (raw.as_str(), *generated))
            .collect();
        assert_eq!(
            models,
            vec![
                ("custom:item/sword", true),
                ("minecraft:item/generated", false),
                ("custom:item/sword_guard", false),
            ]
        );
        assert_eq!(texture_refs.order, vec!["custom:block/steel".to_string()]);
        assert_eq!(blueprint_refs.order, vec!["models/sword".to_string()]);
    }

    #[test]
    fn audits_pack_and_reports_missing_assets() {
        let root = std::env::temp_dir().join(format!("dsh-audit-rs-test-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&root);
        let model_dir = root.join("assets").join("pack").join("models").join("item");
        let items_dir = root.join("assets").join("pack").join("items");
        std::fs::create_dir_all(&model_dir).unwrap();
        std::fs::create_dir_all(&items_dir).unwrap();
        std::fs::write(
            model_dir.join("good.json"),
            r#"{"parent":"minecraft:item/generated","textures":{"layer0":"pack:item/good"}}"#,
        )
        .unwrap();
        std::fs::write(items_dir.join("taken.json"), "{}").unwrap();
        let items: JsonObject = serde_json::from_value(json!({
            "good": {
                "model": { "type": "model", "path": "pack:item/good" },
                "item_model": "pack:taken"
            },
            "bad": { "model": { "type": "model", "path": "pack:item/bad" } }
        }))
        .unwrap();
        let blocks = JsonObject::new();
        let mut diagnostics = DiagnosticBag::new();
        let input = AuditInput {
            resource_root: root.to_string_lossy().into_owned(),
            items: &items,
            blocks: &blocks,
            images: None,
            blueprint_root: None,
        };
        let summary = audit_resource_graph(&input, &mut diagnostics);
        assert_eq!(summary.referenced_models, 2);
        assert_eq!(summary.missing_models, 1);
        // good + its vanilla parent + bad = 3 visited, 1 missing.
        assert_eq!(summary.resolved_models, 2);
        assert_eq!(summary.copied_item_definitions, 1);
        assert_eq!(summary.referenced_textures, 1);
        assert_eq!(summary.missing_textures, 1);
        let _ = std::fs::remove_dir_all(&root);
    }
}
